codalab/lib/bundle_store.py

```diff
@@ -94,74 +81,79 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
         '''
         to_delete = []
 
-        # Create temporary directory as a staging area and put everything there.
-        temp_path = tempfile.mkdtemp('-bundle_store_upload')
-        temp_subpaths = []
+        # If just a single file, set the final path to be equal to that file
+        single_file = len(sources) == 1
+
+        final_path = os.path.join(self.data, uuid)
+        if os.path.exists(final_path):
+            raise UsageError('Error, path %s already present in bundle store' % final_path)
+        # Only make if not there
+        elif not single_file:
+            path_util.make_directory(final_path)
+
+        # Paths to resources
+        subpaths = []
+
         for source in sources:
             # Where to save |source| to (might change this value if we unpack).
-            temp_subpath = os.path.join(temp_path, os.path.basename(source))
+            if not single_file:
+                subpath = os.path.join(final_path, os.path.basename(source))
+            else:
+                subpath = final_path
+
             if remove_sources:
                 to_delete.append(source)
```

Comment:
`to_delete` was only to delete the source in the `temp_dir`, never an original source.  The way it worked before was when you use RemoteBundleClient to upload: you'd copy everything first to a temp_dir, and then call LocalBundelClient.upload_bundle, which would "upload" it to the bundle store with the remove_sources=True, which really meant move it to the `data` directory.

But now, we should just upload directly into the data directory, since we have the uuid (yes!).  This requires modifying RemoteBundleClient.upload_bundle to not use open_temp_file, but to add a command to bundle_rpc_server called open_bundle_for_writing, which (after checking permissions, etc.) allows us to upload directly to the bundle store.

Now that I think about it, it might be good to defer this to another PR since it's going to be somewhat involved...